File: client/src/app/shared/models/product.ts
export type Product = {
  id: number;
  name: string;
  description: string;
  price: number;
  discountedPrice?: number;
  imageUrl: string;
  imageCoverUrls: string[];
  type: string;
  brand: string;
  quantityInStock: number;
  visibility: boolean;
  sku: string;
  barcode: string;
  weight: number;
  category: {
    id: number;
    slug: string;
    status: boolean;
    includeInNav: boolean;
    position: number;
    categoryDescriptions: {
      id: number;
      name: string;
      shortDescription: string;
      description: string;
      image: string;
      urlKey: string;
    }[];
  };
  productAttributes: {
    productAttribute: {
      id: number;
      attributeCode: string;
      attributeName: string;
      type: string;
      isRequired: boolean;
      displayOnFrontend: boolean;
      sortOrder: number;
      isFilterable: boolean;
      attributeOptions: {
        id: number;
        attributeCode: string;
        optionText: string;
        description: null;
      }[];
    };
  }[];
  variants: {
    id: number;
    sku: string;
    price: number;
    imageUrl: string;
    attributeValues: {
      id: number;
      optionText: string;
      variantId: number;
      attributeId: number;
      option: {
        id: number;
        attributeCode: string;
        optionText: string;
        description: null;
      };
    }[];
  }[];
  productCustomOptions: {
    id: number;
    optionName: string;
    optionType: string;
    isRequired: boolean;
    sortOrder: number;
    productCustomOptionValues: {
      id: number;
      extraPrice: number;
      sortOrder: number;
      value: string;
    }[];
  }[];
};

File: server/Services/Product/Core/Entities/Product.cs
namespace Core.Entities;

public class Product : BaseEntity
{
    public required string Name { get; set; }
    public required string Description { get; set; }
    public decimal Price { get; set; }
    public decimal? DiscountedPrice { get; set; }
    public required string ImageUrl { get; set; }
    public List<string> ImageCoverUrls { get; set; }
    public required string Type { get; set; }
    public required string Brand { get; set; }
    public int QuantityInStock { get; set; }

    public bool Visibility { get; set; } = true;
    public required string Sku { get; set; }
    public string? Barcode { get; set; }
    public decimal? Weight { get; set; }
    public Category? Category { get; set; }
    public List<ProductAttributeNavigator> ProductAttributes { get; set; } = new();
    public List<Variant> Variants { get; set; } = new();
    public List<ProductCustomOption> ProductCustomOptions { get; set; } = [];
}
